File: server/db_handler.js
import express from 'express';
import pkg from 'pg';
const { Pool } = pkg;

const router = express.Router();

const pool = new Pool({
  host: process.env.DB_HOST,
  port: process.env.DB_PORT,
  database: process.env.DB_DATABASE,
  user: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  driver: process.env.DB_DRIVER
});

router.get('/years', async (req, res) => {
  const result = await pool.query('SELECT DISTINCT year FROM questions ORDER BY year');
  res.send(result.rows.map(row => row.year));
});

router.get('/exams/:year', async (req, res) => {
  const result = await pool.query('SELECT DISTINCT exam FROM questions WHERE year = $1 ORDER BY exam', [req.params.year]);
  res.send(result.rows.map(row => row.exam));
});

router.get('/parts/:year/:exam', async (req, res) => {
  const result = await pool.query('SELECT DISTINCT part FROM questions WHERE year = $1 AND exam = $2 ORDER BY part', [req.params.year, req.params.exam]);
  res.send(result.rows);
});

router.get('/numbers/:year/:exam/:part', async (req, res) => {
  const result = await pool.query('SELECT DISTINCT number FROM questions WHERE year = $1 AND exam = $2 AND part = $3 ORDER BY number', [req.params.year, req.params.exam, req.params.part]);
  res.send(result.rows);
});

router.get('/questions/:year', async (req, res) => {
    const result = await pool.query('SELECT addr, id FROM questions WHERE year = $1 ORDER BY part, number', [req.params.year]);
    res.send(result.rows);
  });

router.get('/questions/:year/:exam', async (req, res) => {
  const result = await pool.query('SELECT addr, id FROM questions WHERE year = $1 AND exam = $2 ORDER BY part, number', [req.params.year, req.params.exam]);
  res.send(result.rows);
});

router.get('/questions/:year/:exam/:part/:number', async (req, res) => {
  const result = await pool.query('SELECT * FROM questions WHERE year = $1 AND exam = $2 AND part = $3 AND number = $4', [req.params.year, req.params.exam, req.params.part, req.params.number]);
  res.send(result.rows);
});

router.get('/question/:id', async (req, res) => {
  const result = await pool.query('SELECT * FROM questions WHERE id = $1', [req.params.id]);
  if (result.rows.length === 0) {
    res.send({})
  }
  else {
    res.send(result.rows[0]);
  }
});

export default router;

File: server/db_handler.py
import os
import tornado.ioloop
import tornado.web
import tornado.escape
import asyncio
import asyncpg

pool = None

async def make_pool():
    global pool
    pool = await asyncpg.create_pool(
        host=os.getenv("DB_HOST"),
        port=os.getenv("DB_PORT"),
        database=os.getenv("DB_DATABASE"),
        user=os.getenv("DB_USER"),
        password=os.getenv("DB_PASSWORD"),
    )

class BaseHandler(tornado.web.RequestHandler):
    async def prepare(self):
        if not pool:
            await make_pool()

class YearsHandler(BaseHandler):
    async def get(self):
        async with pool.acquire() as connection:
            result = await connection.fetch('SELECT DISTINCT year FROM questions ORDER BY year')
            self.write(tornado.escape.json_encode([row['year'] for row in result]))

class ExamsHandler(BaseHandler):
    async def get(self, year):
        async with pool.acquire() as connection:
            result = await connection.fetch('SELECT DISTINCT exam FROM questions WHERE year = $1 ORDER BY exam', year)
            self.write(tornado.escape.json_encode([row['exam'] for row in result]))

class PartsHandler(BaseHandler):
    async def get(self, year, exam):
        async with pool.acquire() as connection:
            result = await connection.fetch('SELECT DISTINCT part FROM questions WHERE year = $1 AND exam = $2 ORDER BY part', year, exam)
            self.write(tornado.escape.json_encode([row['part'] for row in result]))

class NumbersHandler(BaseHandler):
    async def get(self, year, exam, part):
        async with pool.acquire() as connection:
            result = await connection.fetch('SELECT DISTINCT number FROM questions WHERE year = $1 AND exam = $2 AND part = $3 ORDER BY number', year, exam, part)
            self.write(tornado.escape.json_encode([row['number'] for row in result]))

class QuestionsByYearHandler(BaseHandler):
    async def get(self, year):
        async with pool.acquire() as connection:
            result = await connection.fetch('SELECT addr, id FROM questions WHERE year = $1 ORDER BY part, number', year)
            self.write(tornado.escape.json_encode([dict(row) for row in result]))

class QuestionsByExamHandler(BaseHandler):
    async def get(self, year, exam):
        async with pool.acquire() as connection:
            print(f"year={year}, exam={exam}")
            result = await connection.fetch('SELECT addr, id FROM questions WHERE year = $1 AND exam = $2 ORDER BY part, number', year, exam)
            self.write(tornado.escape.json_encode([dict(row) for row in result]))

class QuestionsHandler(BaseHandler):
    async def get(self, year, exam, part, number):
        async with pool.acquire() as connection:
            result = await connection.fetch('SELECT * FROM questions WHERE year = $1 AND exam = $2 AND part = $3 AND number = $4', year, exam, part, number)
            self.write(tornado.escape.json_encode([dict(row) for row in result]))

class QuestionHandler(BaseHandler):
    async def get(self, id):
        async with pool.acquire() as connection:
            result = await connection.fetch('SELECT * FROM questions WHERE id = $1', id)
            if not result:
                self.write({})
            else:
                self.write(tornado.escape.json_encode(dict(result[0])))
